src/Windows/dll/local_cache.cpp

```diff
@@ -1,33 +1,412 @@
+{
+    throw_if(id.empty(), "The 'id' parameter must not be empty.");
+    throw_if(data_size == 0, "Data cannot be empty.");
+    throw_if(data == nullptr, "Data pointer must not be NULL.");
+
+    init();
+    CacheEntryHeaderV1 header{};
+    header.version = CACHE_V1;
+    header.expiry = expiry;
+
+    std::string filename = get_file_name(id);
+    std::wstring wfilename(filename.begin(), filename.end());
+
+    HANDLE file = OpenHandle(wfilename.c_str(), GENERIC_WRITE, 0, NULL, CREATE_ALWAYS, FILE_ATTRIBUTE_NORMAL, NULL);
+    throw_if(file == INVALID_HANDLE_VALUE, "Open handle failed");
+
+    OVERLAPPED overlapvar = { 0 };
+
+    throw_if(!LockFileEx(file, LOCKFILE_EXCLUSIVE_LOCK, 0, MAXDWORD, MAXDWORD, &overlapvar),
+        "Exclusive write lock operation failed");
+
+    DWORD headerwritten;
+    DWORD datawritten;
+
+    throw_if(!WriteFile(file, &header, sizeof(header), &headerwritten, NULL),
+        "Header write to local cache failed");
+
+    throw_if(!WriteFile(file, data, (DWORD)data_size, &datawritten, NULL),
+        "Data write to local cache failed");
+
+    throw_if(!UnlockFileEx(file, 0, MAXDWORD, MAXDWORD, &overlapvar),
+        "Exclusive write unlock operation failed");
+
+    CloseHandle(file);
 }
 
 std::unique_ptr<std::vector<uint8_t>> local_cache_get(
     const std::string& id)
 {
-    (void)id;
-    return nullptr;
+    throw_if(id.empty(), "The 'id' parameter must not be empty.");
+    init();
+
+    std::string filename = get_file_name(id);
+    std::wstring wfilename(filename.begin(), filename.end());
+    
+    HANDLE file = OpenHandle(wfilename.c_str(), GENERIC_READ, FILE_SHARE_READ, NULL, OPEN_EXISTING, FILE_ATTRIBUTE_NORMAL, NULL);
+    if (file == INVALID_HANDLE_VALUE)
+    {
+        return nullptr;
+    }
+
+    OVERLAPPED overlapvar = { 0 };
+
+    throw_if(!LockFileEx(file, 0, 0, MAXDWORD, MAXDWORD, &overlapvar),
+        "Shared read lock operation failed");
+
+    CacheEntryHeaderV1 *header;
+    char buf[sizeof(CacheEntryHeaderV1)] = { 0 };
+    LPDWORD headerread = 0;
+
+    throw_if(!ReadFile(file, &buf, sizeof(CacheEntryHeaderV1), headerread, NULL),
+        "Header read from local cache failed");
+
+    header = (CacheEntryHeaderV1*)buf;
+
+    if (header->expiry <= time(nullptr))
+    {
+        throw_if(!UnlockFileEx(file, 0, MAXDWORD, MAXDWORD, &overlapvar),
+            "Shared read unlock operation failed");
+
+        CloseHandle(file);
+        DeleteFileW(wfilename.c_str());
+        // Even if unlink fails, we can just return null. Thus, the return
+        // value is intentionally ignored here.
+        return nullptr;
+    }
+
+    DWORD size = GetFileSize(file, NULL);
+    int datasize = size - sizeof(CacheEntryHeaderV1);
+    auto cache_entry = std::make_unique<std::vector<uint8_t>>(datasize);
+
+    LPDWORD dataread = 0;
+    
+    throw_if(!ReadFile(file, cache_entry->data(), size, dataread, NULL),
```

Comment:
What happens if the read returns less than datasize?

Reply:
Added in a separate check for this.